staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
 	// Optional: Defaults to true.
 	// +optional
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
+	// domains. Scheduler will schedule pods in a way which abides by the constraints.
+	// This field is alpha-level and is only honored by clusters that enables the EvenPodsSpread
+	// feature.
+	// +optional
+	// +patchMergeKey=topologyKey
+	// +patchStrategy=merge
+	TopologySpreadConstraints []TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty" patchStrategy:"merge" patchMergeKey:"topologyKey" protobuf:"bytes,31,opt,name=topologySpreadConstraints"`
+}
+
+type UnsatisfiableConstraintAction string
+
+const (
+	// DoNotSchedule instructs the scheduler not to schedule the pod
+	// when constraints are not satisfied.
+	DoNotSchedule UnsatisfiableConstraintAction = "DoNotSchedule"
+	// ScheduleAnyway instructs the scheduler to schedule the pod
+	// even if constraints are not satisfied.
+	ScheduleAnyway UnsatisfiableConstraintAction = "ScheduleAnyway"
+)
+
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree to which pods may be unevenly distributed.
+	// It's the maximum permitted difference between the number of matching pods in
+	// any two topology domains of a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// are "spread" such that zone1 and zone2 each have one pod, but not zone3.
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// scheduling it onto zone1(zone2) would make the ActualSkew(2) violate MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required field. Default value is 1 and 0 is not allowed.
+	MaxSkew int32 `json:"maxSkew" protobuf:"varint,1,opt,name=maxSkew"`
+	// TopologyKey is the key of node labels. Nodes that have a label with this key
+	// and identical values are considered to be in the same topology.
+	// We consider each <key, value> as a "bucket", and try to put balanced number
+	// of pods into each bucket.
+	// It's a required field.
+	TopologyKey string `json:"topologyKey" protobuf:"bytes,2,opt,name=topologyKey"`
+	// WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+	// the spread constraint.
+	// - DoNotSchedule (default) tells the scheduler not to schedule it
+	// - ScheduleAnyway tells the scheduler to still schedule it
+	// Note: it's considered as "Unsatisfiable" only when actual skew on all nodes
+	// exceeds "MaxSkew".
```

Comment:
...and this condition can't be fixed or improved by adding an additional pod?

e.g. suppose max skew is 1, and the number of pods across 3 topologies is 3/3/1. It seems we should be able to schedule because that would fix the imbalance, right? What if the split is 3/1/1? Then adding 2 pods could fix it. If there are two pending pods, does the system not schedule either of them?

Reply:
In both the 2 cases you gave, we don't see pending pods (unless due to failure on other predicates like node resource shortage). For the split 3/3/1, the incoming pod can be put to topology 3; for the split 3/1/1, the incoming pod can be put to topology 2/3.

One example we will see pending pod is:

```
+---------------+-------+
|     zone1     | zone2 |
+---------------+-------+
| nodeA | nodeB | nodeC |
+-------+-------+-------+
|  P P  |   P   |  P P  |
+-------+-------+-------+
```

Suppose we have two topologySpreadConstraints on both zone and node, with maxSkew 1. In the above example, incoming pod can't be placed:

- if it's put to zone1, the split on **zone** would be imbalanced: 4 vs. 2
- if it's put to zone2(nodeC), the split on **node** would be imbalanced: 2 vs. **1** vs. **3**

Reply:
Please update the comment.

Reply:
> it's not clear that this permits assignments which improve the skew when it is out of the allowed range

Not exactly. It doesn't improve the skew - it's the "minMatch" that got updated.
For the 3/3/1 case, "minMatch" was 1:
- if we put the incoming pod to topology 3, the global "minMatch" becomes 2, this is allowed.
- if we put the incoming pod to topology 1/2, the global "minMatch" remains 1, so for sure it's not allowed.

Reply:
My actual question is about the case e.g. 4/4/1. If you have two pods to schedule, I think the behavior people will want is that they both schedule and you end up with 4/4/3. I think the behavior you're describing is that neither pod will schedule. Please make the behavior of this condition clear in the comment.

Also let me urge you to make it permit scheduling pods that don't make things worse: imagine 4/1/1 with 6 pending pods to schedule. IMO the system is unacceptably brittle if the scheduler can't fix things by scheduling the pending pods.

I'm going OOO so I won't be able to continue this conversation, very sorry.

Reply:
> Please make the behavior of this condition clear in the comment.

Will do.

> Also let me urge you to make it permit scheduling pods that don't make things worse: imagine 4/1/1 with 6 pending pods to schedule. IMO the system is unacceptably brittle if the scheduler can't fix things by scheduling the pending pods.

Exactly. The internal implementation is exactly as what you said - the only possible final result is from 4/1/1 to 4/4/4.

> I'm going OOO so I won't be able to continue this conversation, very sorry.

It's totally fine as we will target this to 1.16. And during your absence, I will request reviewers to review afterward PRs which are only related with scheduling.

Appreciate your efforts reviewing this!

Reply:
Comments updated.